pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,300 @@
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+	FlowSchemaConditionTrue    FlowSchemaConditionStatus = "True"
+	FlowSchemaConditionFalse   FlowSchemaConditionStatus = "False"
+	FlowSchemaConditionUnknown FlowSchemaConditionStatus = "Unknown"
+
+	PriorityLevelConfigurationConditionTrue    PriorityLevelConfigurationConditionStatus = "True"
+	PriorityLevelConfigurationConditionFalse   PriorityLevelConfigurationConditionStatus = "False"
+	PriorityLevelConfigurationConditionUnknown PriorityLevelConfigurationConditionStatus = "Unknown"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchema defines the schema of a group of flows. Note that a flow makes up of a set of inbound API requests with
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// Current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// PriorityLevelConfigurationRef can reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will ignored and marked as invalid in its status.
+	PriorityLevelConfigurationRef PriorityLevelConfigurationRef
+	// MatchingPriority controls the ordering the flow-schemas in the cluster. Its value must be greater than zero. The
+	// primary sort is based on MatchingPriority, ordered highest to lowest (20 before 10).
+	MatchingPriority int32
+	// Distinguisher defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	// +optional
+	Distinguisher *FlowDistinguisher
+	// Rules holds all the RuleWithUser for this FlowSchema.
+	Rules []RuleWithUser
+}
+
+// FlowDistinguisher specifies the configuration of the distinguisher.
+type FlowDistinguisher struct {
```

Comment:
This is not a flow distinguisher, rather it is a prescription for how to compute a flow distinguisher.  Perhaps we should call this `FlowDistinguisherSpec` or `FlowDistinguisherRecipe`?

Reply:
the name `FlowDistinguisherSpec` will mess up w/ the standard `spec` of kubernetes objects. `FlowDistinguisher`

> what about `FlowDistinguisherType` here? (or any shorter name?)

makes it a string-alias named `FlowDistinguisherType`?

Reply:
There is nothing magic in "Spec".  Using `FlowDistinguisherSpec` for this would only be a problem if we wanted a top-level object type named `FlowDistinguisher` (which we do not).